### PR TITLE
DeepSpeed sequence parallelism (aka Ulysses) integration with HF transformer

### PR DESCRIPTION
**Pull request by samadejacobs**
# What does this PR do?
This PR enhances capabilities of [DeepSpeed long sequence (context) parallelism (aka DS Ulysses)](https://dl.acm.org/doi/10.1145/3662158.3662806) with support for HF models. Support is currently enabled when both DeepSpeed and flash attn are enabled. Future support would be extended to SPDA.  All current and future HF models (such as Llama, opt etc) using refactored flash_attention_utils are supported.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -20,9 +20,18 @@
...
 
+from .integrations.deepspeed import (  # DeepSpeed seq parallelism (aka Ulysses)
+    is_deepspeed_available,
+    is_deepspeed_sp_enabled,
+)
 from .utils import is_flash_attn_2_available, is_flash_attn_greater_or_equal
 
 
+if is_deepspeed_available():
+    from deepspeed.sequence.layer import _SeqAllToAll
```
Comment (muellerzr) [inline src/transformers/modeling_flash_attention_utils.py:31]: Not sure I enjoy the fact that we rely on the private `_SeqAllToAll`, anything we should worry about there in the future?

File: tests/deepspeed/test_ulysses.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import sys
+
+import torch
+from deepspeed import initialize
```
Comment (muellerzr) [inline tests/deepspeed/test_ulysses.py:19]: These both need to be guarded under their respective `if_xxx_available`, it's why the current tests are failing

File: docs/source/en/deepspeed.md [hunk 3]
```diff
@@ -1141,6 +1141,52 @@ Using multiple GPUs with ZeRO-3 for generation requires synchronizing the GPUs b
...
+
+for n, batch in enumerate(data_loader):
+    seq_length = batch["input_ids"].size(1)
+    assert seq_length % seq_parallel_world_size == 0
+    sub_seq_length = seq_length // seq_parallel_world_size
+    sub_seq_start = seq_parallel_rank * sub_seq_length
+    sub_seq_end = (seq_parallel_rank + 1) * sub_seq_length
+
+    batch["input_ids"] = batch["input_ids"][:, sub_seq_start:sub_seq_end]
+    batch["labels"] = batch["labels"][:, sub_seq_start:sub_seq_end]
```
Comment (muellerzr) [inline docs/source/en/deepspeed.md:1179]: Now that we have this in the Trainer, it'd be good if we can accelerate do this via it's init in the config + we advertise it here. 

It'd be even *better* if we can modify the dataloaders to do the sequence parallel batching automatically, though I know this first was a transformers PR :)

Comment (muellerzr) [reply docs/source/en/deepspeed.md:1179]: Perhaps rn we include a "accelerate version coming soon"?

**Review by muellerzr — APPROVED** [review]
Thanks! This looks great to me, well done. Made a few comments/questions, next let's get the accelerate version up and running 🔥

File: tests/deepspeed/test_ulysses.py (hunk 2 above)
Comment (samadejacobs) [reply tests/deepspeed/test_ulysses.py:19]: The test (in top-level test_deepspeed.py) is already guarded `@required_deepspeed` and [@required_torch_accelerator](https://github.com/huggingface/transformers/blob/9ea2571e79729616c86a6e5dc0ab942a8f4447fc/tests/deepspeed/test_deepspeed.py#L1339). I am thinking that should take of it, no?

File: src/transformers/modeling_flash_attention_utils.py (hunk 1 above)
Comment (samadejacobs) [reply src/transformers/modeling_flash_attention_utils.py:31]: Nothing to worry about!

File: tests/deepspeed/test_ulysses.py (hunk 2 above)
Comment (muellerzr) [reply tests/deepspeed/test_ulysses.py:19]: Nope as it's being imported, we need to guard it from regular import checks.

File: src/transformers/modeling_flash_attention_utils.py [hunk 4]
```diff
@@ -20,9 +20,18 @@
...
+    from deepspeed.utils import groups as ds_comm_groups
```
Comment (SunMarc) [inline src/transformers/modeling_flash_attention_utils.py:32]: Can we only import these function in the block where we use them. We don't want the user to implicitly import them when they don't use it + It will make transformers import slower.

**Review by SunMarc — APPROVED** [review]
Nice, thanks for adding this @samadejacobs  ! Just a nit

File: src/transformers/modeling_flash_attention_utils.py (hunk 4 above)
Comment (samadejacobs) [reply src/transformers/modeling_flash_attention_utils.py:32]: Thanks for pointing this out @SunMarc, I was torn between placing import at the top or within the block where it is needed. I decided to go with the former for performance reason. Recall that _flash_attention_forward_ would be called multiple times, this could lead to repeated imports which could (possibly) degrade performance. Code readerability and mantainability are the other (minor) reasons for my choice.

Comment (SunMarc) [reply src/transformers/modeling_flash_attention_utils.py:32]: Oh indeed ! Thx for explaining. A potential solution could be to import under the condition if is_deepspeed_sp_enabled() instead since we are only using these imports with that condition. WDTY ? Could you add a quick comment on why we decided to put the import at the top ?

File: src/transformers/modeling_flash_attention_utils.py [hunk 5]
```diff
@@ -220,6 +229,16 @@ def _flash_attention_forward(
...
     """
+    if is_deepspeed_sp_enabled():
+        spg = ds_comm_groups._get_sequence_parallel_group()
+        # qkv tensors are of shape (batch_size, seq_len, num_heads, head_dim)
+        scatter_idx = 2  # Scatter on num_heads dimension
+        gather_idx = 1  # Gather on seq_len dimension
+        batch_dim_idx = 0  # Synonymous with the batch_first==true
+        query_states = _SeqAllToAll.apply(spg, query_states, scatter_idx, gather_idx, batch_dim_idx)
+        key_states = _SeqAllToAll.apply(spg, key_states, scatter_idx, gather_idx, batch_dim_idx)
+        value_states = _SeqAllToAll.apply(spg, value_states, scatter_idx, gather_idx, batch_dim_idx)
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:240]: I am a tad bit unconvinced by checking this for each model that runs flash attention. 
We could: 
- either have self.use_deep_speed, and check for that instead
- separate the deepspeed code, by change the forward if deepspeed is available with deepspeed forward. 
For such changes it is important to make sure we don't introduce regressions in the perfromances. For a single forward pass it's gonna small, but for generation, it might add up!

File: docs/source/en/deepspeed.md [hunk 6]
```diff
@@ -1141,6 +1141,52 @@ Using multiple GPUs with ZeRO-3 for generation requires synchronizing the GPUs b
 
 For Transformers>=4.28, if `synced_gpus` is automatically set to `True` if multiple GPUs are detected during generation.
 
+### Non-Trainer Sequence Parallelism
+DeepSpeed sequence parallelism, also known as [DeepSpeed Ulysses](https://github.com/microsoft/DeepSpeed/blob/master/blogs/deepspeed-ulysses/README.md), is compatible with HuggingFace Transformers by adding 'sequence_parallel_size' and 'data_parallel_size' to the DeepSpeed configuration. Additionally, it's required that the user’s script correctly shard the input data along the sequence dimension.
```
Comment (ArthurZucker) [inline docs/source/en/deepspeed.md:1145]: Can we also explain why anyone would want to use this?

**Review by ArthurZucker — COMMENTED** [review]
Thanks, let's make sure this does not impact any other models in terms of speed!

File: src/transformers/modeling_flash_attention_utils.py (hunk 5 above)
Comment (samadejacobs) [reply src/transformers/modeling_flash_attention_utils.py:240]: @ArthurZucker, as to your first bullet point, included in `is_deepspeed_sp_enabled` is the check for "use_deep_speed". Generally speaking, I see your point of view but am also concern about duplicated code (your second bullet point).  Is bullet point 1 without burden on user to set another `use_deep_speed` flag sufficient for you? From deepspeed persepective, we assume that if the user enables deepspeed and sets `sp_size > 1` during initialization, it means they intend to use `deepspeed_sp`.

File: docs/source/en/deepspeed.md (hunk 6 above)
Comment (ArthurZucker) [reply docs/source/en/deepspeed.md:1145]: (I don't mean this as agresssive! More like "if you want to do xxx or yyy this is for you! vs deepspeed normal!)

File: src/transformers/modeling_flash_attention_utils.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/modeling_flash_attention_utils.py:240]: >  if the user enables deepspeed and sets sp_size > 1 during initialization, it means they intend to use deepspeed_sp 

yeah, IMO this is enough!

File: docs/source/en/deepspeed.md (hunk 6 above)
Comment (samadejacobs) [reply docs/source/en/deepspeed.md:1145]: @ArthurZucker, it comes down to user preference. We’ve noticed users utilizing both options from the long list of issues in the DeepSpeed repo. Additionally, there is another PR (number 2877, I think) in progress to support the Accelerate version. At the time this PR was submitted, it was agreed that the pending PR 2877 would leverage this PR to provide support for the Accelerate trainer version. cc: @muellerzr

**Review by ArthurZucker — DISMISSED** [review]
Good to merge I just wanted docs to explain a bit more benefits ( when are perf better with this!)

**Comment by glowwormX** [comment]
@samadejacobs  I'm glad to see this pr will be merged soon. When are you going to support sdpa in the future? It's useful for me.

File: src/transformers/modeling_flash_attention_utils.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/modeling_flash_attention_utils.py:240]: WDYT?

**Comment by glowwormX** [comment]
> @samadejacobs I'm glad to see this pr will be merged soon. When are you going to support sdpa in the future? It's useful for me.

Yes, but I want to run it on the npu, but it doesn't support flash2, but sdpa.

File: docs/source/en/deepspeed.md (hunk 6 above)
Comment (ldh127) [reply docs/source/en/deepspeed.md:1145]: > 我们还可以解释一下为什么有人想使用它吗？

it is very import , shy donot we not need ? for long context train , it is needs

**Comment by ArthurZucker** [comment]
@samadejacobs anything I can do to help get this merged?

File: src/transformers/modeling_flash_attention_utils.py (hunk 5 above)
Comment (samadejacobs) [reply src/transformers/modeling_flash_attention_utils.py:240]: @ArthurZucker, apologies for late response, was OOF! Your suggestion LGTM, please feel to update this PR (or refactor as a follow up PR).

**Comment by samadejacobs** [comment]
> @samadejacobs I'm glad to see this pr will be merged soon. When are you going to support sdpa in the future? It's useful for me.

@glowwormX, future support would be extended to SPDA.

**Comment by ArthurZucker** [comment]
Hey @samadejacobs ! 
Ah, I am not sure I have bandwidth right now for https://github.com/huggingface/transformers/pull/32305#discussion_r1769601049, but if you can't do it will see if I can ping someone or do it!

**Comment by jyshee** [comment]
@samadejacobs so how is this going now?

**Comment by ronald-d-rogers** [comment]
I have been working on this for my team in Bloomberg. I think I may have a PR that merges changes from everybody that works. Will share shortly and hopefully we can get this feature in.

**Comment by SunMarc** [comment]
cc @XuehaiPan

**Comment by pavelgein** [comment]
Hi, I'd like to know about plan to merge it and I would be happy to help with it.

I have recently [implemented](https://github.com/turbo-llm/turbo-alignment/pull/67) Ulyssys attention in our turbo-alignment library, which is built on Transformers library and heavily uses both Transformers models and Trainer. 
Despite not being merged, that PR is quite well tested and we have sucesfully trained several models with large context with this code.

From this experience I have learnt, that to achieve  this, we need to take several steps.

1. Add or patch existing attention modules. Like in this PR, this could be done with a decorator. The problem is eager attention module, which is repeated across many models, [Gemma](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma2/modular_gemma2.py#L216) and [Llama](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L171) for example. It can be questioned whether one, who would like to use context parallelism, will use eager implementation, so IMHO, eager attention can be left for later.

2. Adjust all computations which use world size. For example, number of training steps in the Trainer [code](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2277), and in some places in the Accelerate, for example in data loader [preparing](https://github.com/huggingface/accelerate/blob/main/src/accelerate/accelerator.py#L2339). This is important for right datasets sharding across workers.

3. Perfectly synchronize all the random generators inside one sequence parallel group, since otherwise, one can get different permutations 

4. Deal with position numbering and KV caches. As of today, in the absence of cache, positions_ids will be assigned to numbers from 0 to `seq_length - 1` and this has to be adjusted to reflect that i-th worker in the sequence parallel group see only subset of tokens (see Gemma 2 [code](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma2/modular_gemma2.py#L413-L420) for example. This is not hard to do. But I don't know how to deal with caches, in our implementation we explicitly disable them. 

5. Attention masks. It seem right, that each worker inside sequence parallel group sees all the attention mask (worker needs it to compute the attention), so some around computing lengths and cache positions should be adjusted. 

6. Generation. The generation process requires that we at each step regroup the whole sentence. Also, at each step we have to deal with KV caches

7. Workers discovery. As far as I know, there are two approaches: model parallel unit and torch device mesh.


I would love to discuss this issuses and help to deal with them.

I am also aware of [another](https://github.com/huggingface/transformers/issues/35983) context parallel approach, but seems like this great library can have both, right?

**Comment by SunMarc** [comment]
We will soon have a new PR from deepspeed team ! I will close this one once they open it. Sorry that this one didn't get merged.

**Comment by ciuffredaluca** [comment]
> We will soon have a new PR from deepspeed team ! I will close this one once they open it. Sorry that this one didn't get merged.

Hi @SunMarc, is there any update on this topic? Any expected timeline? Thanks :)

**Comment by SunMarc** [comment]
cc @stas00 do you know when you are planning to land this ?

**Comment by stas00** [comment]
@SunMarc, please kindly sync with @S1ro1 - we are waiting for him to complete the redesign of the parallelism in HF Accelerate. https://github.com/huggingface/accelerate/pull/3673

**Comment by S1ro1** [comment]
@stas00 I think given my limited availability recently and the time I'll be able to get to doing it in DeepSpeed, you can probably just integrate it with DeepSpeed as is and we'll do the refactor separately, else I don't think its entirely realistic to get it merged soon enough.

**Comment by stas00** [comment]
> you can probably just integrate it with DeepSpeed

I'm not sure what you mean, Matej. It's already in Deepspeed. Unless you mean in the Deepspeed plugin of HF Accelerate?

What's the rough new ETA for your work?
